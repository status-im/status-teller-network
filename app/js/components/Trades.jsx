import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Row, Col, Card } from 'reactstrap';
import { Link } from "react-router-dom";
import { withNamespaces } from 'react-i18next';
import Blockies from 'react-blockies';
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faCircle } from "@fortawesome/free-solid-svg-icons";

class Trades extends Component {
  renderTrades() {
    return (
      <Card body className="py-0">
        {this.props.trades.map((trade, index) => (
          <div key={index} className="d-flex my-1">
            <span className="flex-fill align-self-center">
              <Blockies seed={trade.address} className="align-middle rounded-circle"/>
              <span className="ml-2">{trade.name}</span>
            </span>
            <span className="flex-fill align-self-center">{trade.value} ETH</span>
            <span className="flex-fill align-self-center text-right text-success">
              <FontAwesomeIcon icon={faCircle} className="mr-2"/>
              {trade.status}
            </span>
          </div>
        ))}
      </Card>
    );
  }

  renderEmpty(t) {
    return (
      <Card body className="text-center">
        {t('trades.noOpen')}
      </Card>
    );
  }

  render() {
    const t = this.props.t;
    return (
      <Row className="mt-4">
        <Col xs="12">
          <span className="font-weight-bold h5">{t('trades.title')}</span>
          <Link to="/buy" className="float-right">{t('trades.find')}</Link>
        </Col>
        <Col xs="12">
          {this.props.trades.length === 0 ? this.renderEmpty(t) : this.renderTrades()}
        </Col>
      </Row>
    );
  }
}

Trades.propTypes = {
  t: PropTypes.func,
  trades: PropTypes.array
};

export default withNamespaces()(Trades);
