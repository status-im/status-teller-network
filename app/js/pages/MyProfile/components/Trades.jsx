import React, {Component} from 'react';
import PropTypes from 'prop-types';
import {Card} from 'reactstrap';
import {Link} from "react-router-dom";
import {withNamespaces} from 'react-i18next';
import {FontAwesomeIcon} from "@fortawesome/react-fontawesome";
import {faCircle, faArrowRight} from "@fortawesome/free-solid-svg-icons";
import Identicon from "../../../components/UserInformation/Identicon";

class Trades extends Component {
  renderTrades() {
    return (
      <Card body className="py-2 px-3 shadow-sm">
        {this.props.trades.map((trade, index) => (
          <div key={index} className="d-flex my-1">
            <span className="flex-fill align-self-center">
              <Identicon seed={trade.offer.owner} className="align-middle rounded-circle border"/>
              <span className="ml-2">{trade.offer.user.username}</span>
            </span>
            <span className="flex-fill align-self-center">{trade.amount} {trade.offer.token.symbol}</span>
            <span className="flex-fill align-self-center text-right text-success">
              <FontAwesomeIcon icon={faCircle} className="mr-2"/>
              {trade.status}
            </span>
          </div>
        ))}
      </Card>
    );
  }

  renderEmpty() {
    const {t} = this.props;
    return (
      <Card body className="text-center">
        {t('trades.noOpen')}
      </Card>
    );
  }

  render() {
    const {t, trades} = this.props;
    return (
      <div className="mt-3">
        <div>
          <h3 className="d-inline-block">{t('trades.title')}</h3>
          <span className="float-right">
            <Link to="/buy" className="float-right">{t('trades.find')} <FontAwesomeIcon icon={faArrowRight}/></Link>
          </span>
        </div>
        {trades.length === 0 ? this.renderEmpty(t) : this.renderTrades()}
      </div>
    );
  }
}

Trades.propTypes = {
  t: PropTypes.func,
  trades: PropTypes.array
};

export default withNamespaces()(Trades);
