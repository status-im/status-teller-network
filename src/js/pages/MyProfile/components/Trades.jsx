import React, {Component} from 'react';
import PropTypes from 'prop-types';
import {Card, Row, Col} from 'reactstrap';
import {Link} from "react-router-dom";
import {withNamespaces} from 'react-i18next';
import {FontAwesomeIcon} from "@fortawesome/react-fontawesome";
import {faArrowRight} from "@fortawesome/free-solid-svg-icons";
import Identicon from "../../../components/UserInformation/Identicon";
import {formatBalance} from "../../../utils/numbers";
import {tradeStates} from "../../../features/escrow/helpers";

const getTradeStyle = (trade) => {
  if(trade.arbitration){
    if(trade.arbitration.open){
      trade.status = tradeStates.arbitration_open;
    } else {
      if(trade.arbitration.result !== '0'){
        trade.status = tradeStates.arbitration_closed;
      }
    }
  }

  switch(trade.status){
    case tradeStates.waiting:
    case tradeStates.funded:
    case tradeStates.paid:
      return {text: 'Open', className: 'bg-success'};
    case tradeStates.released:
      return {text: 'Closed', className: 'bg-primary'};
    case tradeStates.canceled:
      return {text: 'Canceled', className: 'bg-secondary'};
    case tradeStates.expired:
      return {text: 'Expired', className: 'bg-secondary'};
    case tradeStates.arbitration_open:
      return {text: 'Arbitration', className: 'bg-danger'};
    case tradeStates.arbitration_closed:
      return {text: 'Resolved', className: 'bg-primary'};
    default:
      return {text: trade.status, className: 'bg-secondary'};
  }
};

class Trades extends Component {
  renderTrades() {
    const address = this.props.address;
    return (
      <Card body className="py-2 px-3 shadow-sm">
        {this.props.trades.map((trade, index) => {
          const isBuyer = trade.buyer ===  address;
          const tradeStyle = getTradeStyle(trade);
          return <Link key={index} to={"/escrow/" + trade.escrowId}>
            <Row className="my-1 border-bottom">
              <Col className="align-self-center pr-0" xs="2">
                <Identicon seed={ isBuyer ? trade.offer.owner : trade.buyer} scale={5} className="align-middle rounded-circle topCircle border"/>
              </Col>
              <Col className="align-self-center" xs="3">
                <span>{isBuyer ? trade.seller.username : trade.buyerInfo.username}</span>
              </Col>
              <Col className="align-self-center" xs="3">
                {isBuyer ? 'Buy' : 'Sell' } {formatBalance(trade.tokenAmount)} {trade.token.symbol}
              </Col>
              <Col className="align-self-center text-center text-success" xs="4">
                <span className={"p-1 text-uppercase d-inline text-white rounded-sm text-small " + tradeStyle.className}>{tradeStyle.text}</span>
              </Col>
            </Row>
          </Link>;
        })}
      </Card>
    );
  }

  renderEmpty() {
    const {t} = this.props;
    return (
      <Card body className="text-center">
        {t('trades.noOpen')}
      </Card>
    );
  }

  render() {
    const {t, trades} = this.props;
    return (
      <div className="mt-3">
        <div>
          <h3 className="d-inline-block">{t('trades.title')}</h3>
          <span className="float-right">
            <Link to="/buy" className="float-right">{t('trades.find')} <FontAwesomeIcon icon={faArrowRight}/></Link>
          </span>
        </div>
        {trades.length === 0 ? this.renderEmpty(t) : this.renderTrades()}
      </div>
    );
  }
}

Trades.propTypes = {
  t: PropTypes.func,
  trades: PropTypes.array,
  address: PropTypes.string
};

export default withNamespaces()(Trades);
