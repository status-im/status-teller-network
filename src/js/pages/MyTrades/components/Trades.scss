@import "../../../../css/variable-overrides";

.state-filters {
  .form-control {
    height: 40px;
  }
}