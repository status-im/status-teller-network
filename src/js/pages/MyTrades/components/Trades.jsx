import React, {Component, Fragment} from 'react';
import PropTypes from 'prop-types';
import {Card, Row, Col, Form, FormGroup, Label, Input} from 'reactstrap';
import {Link} from "react-router-dom";
import {withNamespaces} from 'react-i18next';
import {FontAwesomeIcon} from "@fortawesome/react-fontawesome";
import {faCaretDown} from "@fortawesome/free-solid-svg-icons";
import Identicon from "../../../components/UserInformation/Identicon";
import {formatBalance} from "../../../utils/numbers";
import {tradeStates, tradeStatesFormatted, completedStates} from "../../../features/escrow/helpers";
import {addressCompare} from "../../../utils/address";
import {ARBITRATION_SOLVED_BUYER, ARBITRATION_SOLVED_SELLER} from "../../../features/arbitration/constants";

import './Trades.scss';

const getTradeStyle = (trade, isBuyer) => {
  if (trade.mining) {
    return {text: 'Mining', className: 'bg-info'};
  }
  if(trade.arbitration){
    if(trade.arbitration.open){
      trade.status = tradeStates.arbitration_open;
    } else {
      if(trade.arbitration.result !== '0'){
        trade.status = tradeStates.arbitration_closed;
      }
    }
  }
  const tradeStyle = {text: tradeStatesFormatted[trade.status]};

  switch(trade.status){
    case tradeStates.waiting: tradeStyle.className = 'bg-primary'; break;
    case tradeStates.funded: tradeStyle.className = 'bg-primary'; break;
    case tradeStates.paid: tradeStyle.className = 'bg-primary'; break;
    case tradeStates.released: tradeStyle.className = 'bg-success'; break;
    case tradeStates.canceled: tradeStyle.className = 'bg-secondary text-black'; break;
    case tradeStates.expired: tradeStyle.className = 'bg-secondary text-black'; break;
    case tradeStates.arbitration_open: tradeStyle.className = 'bg-warning'; break;
    case tradeStates.arbitration_closed: {
      let className;
      if (trade.arbitration.result.toString() === ARBITRATION_SOLVED_BUYER) {
        className = isBuyer ? 'bg-success' : 'bg-danger';
      } else if (trade.arbitration.result.toString() === ARBITRATION_SOLVED_SELLER) {
        className = !isBuyer ? 'bg-success' : 'bg-danger';
      } else {
        className = 'bg-primary';
      }
      tradeStyle.className = className; break;
    }
    default: tradeStyle.className = 'bg-secondary text-black'; break;
  }
  return tradeStyle;
};

class Trades extends Component {
  state = {
    filteredState: '',
    showFilters: false
  };

  filterState(stateName) {
    if (stateName === '') {
      return this.setState({filteredState: ''});
    }
    const stateIndex = Object.values(tradeStatesFormatted).findIndex(tradeState => tradeState === stateName);
    const filteredState = stateIndex === -1 ? '' : Object.keys(tradeStatesFormatted)[stateIndex];
    this.setState({filteredState});
  }

  renderTrades() {
    const address = this.props.address;
    return (
      <Fragment>

        {this.state.showFilters && <Form><b>Filters:</b>
          <FormGroup row className="state-filters">
            <Label for="trade-state-filter" xs={2}>States</Label>
            <Col xs={10}>
              <Input type="select" name="select" id="trade-state-filter"
                     onChange={(e) => this.filterState(e.target.value)}>
                <option/>
                { Object.keys(tradeStatesFormatted)
                        .filter(x => { return this.props.active ?  !completedStates.includes(x) : completedStates.includes(x); })
                        .map((tradeState, index) => <option
                  key={`tradeState-${index}`}>{tradeStatesFormatted[tradeState]}</option>)}
              </Input>
            </Col>
          </FormGroup>
        </Form>}

        {!this.state.showFilters && <p className="text-small text-muted mb-1 clickable" onClick={() => this.setState({showFilters: true})}>Show Filters <FontAwesomeIcon icon={faCaretDown}/></p>}

        <Card body className="profile-trades-list border-0 pt-0">
          {(() => {
            const trades = this.props.trades.filter(x => { return this.props.active ? !completedStates.includes(x.status) : completedStates.includes(x.status); }).map((trade, index) => {
              if (this.state.filteredState && trade.status !== this.state.filteredState) {
                return null;
              }
              const isBuyer = addressCompare(trade.buyer, address);
              const tradeStyle = getTradeStyle(trade, isBuyer);
              return <Link key={index} to={"/escrow/" + trade.escrowId} className="text-black">
                <Row className="my-1 border-bottom shadow-sm p-2 mb-3">
                  <Col className="align-self-center pr-0" xs="2">
                    <Identicon seed={isBuyer ? trade.seller.statusContactCode : trade.buyerInfo.statusContactCode}
                               scale={5} className="align-middle rounded-circle topCircle border"/>
                  </Col>
                  <Col className="align-self-center" xs="3">
                    <span>{isBuyer ? trade.seller.username : trade.buyerInfo.username}</span>
                  </Col>
                  <Col className="align-self-center" xs="3" md={4}>
                    {isBuyer ? 'Buy' : 'Sell'} {formatBalance(trade.tokenAmount)} {trade.token.symbol}
                  </Col>
                  <Col className="align-self-center text-center text-success" xs="4" md={3}>
                  <span
                    className={"p-1 text-uppercase d-inline text-white rounded-sm text-small nowrap " + tradeStyle.className}>{tradeStyle.text}</span>
                  </Col>
                </Row>
              </Link>;
            });
            if (trades.every(trade => trade === null)) {
              return (
                <Row className="my-1 border-bottom shadow-sm p-2 mb-3">
                  <Col className="align-self-center pt-3 pb-3 text-center text-muted">{this.props.t('trades.noFilteredTrades')}</Col>
                </Row>
              )
            }
            return trades;
          })()}
        </Card>
      </Fragment>
    );
  }

  renderEmpty() {
    const {t, active} = this.props;
    return (
      <Card body className="text-center border-0 text-muted shadow-sm">
        {active ? t('trades.noActiveTrades') : t('trades.noPastTrades')}
      </Card>
    );
  }

  render() {
    const {t, trades} = this.props;
    return (
      <div className="mb-5">
        {trades.length === 0 ? this.renderEmpty(t) : this.renderTrades()}
      </div>
    );
  }
}

Trades.defaultProps = {
  active: false
};

Trades.propTypes = {
  t: PropTypes.func,
  trades: PropTypes.array,
  address: PropTypes.string,
  active: PropTypes.bool
};

export default withNamespaces()(Trades);
