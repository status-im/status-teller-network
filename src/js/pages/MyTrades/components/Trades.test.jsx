import React from "react";
import { shallow } from "enzyme";

import Trades from "./Trades";

/* eslint-disable-next-line no-unused-vars*/
global.web3 = { 
  utils: {
    toChecksumAddress: () => true
  }
 };

describe('Trades', () => {
  it('should render correctly', () => {
    const component = shallow(<Trades trades={[{buyerInfo: {username: 'Name'}, seller: {username: 'Name'},  token: { symbol: 'SNT' }, offer: { owner: '0x0', user: { username: 'Name' } }, value: '2', status: 'open', escrowId: 0}]} />);
  
    expect(component).toMatchSnapshot();
  });

  it('should render when empty', () => {
    const component = shallow(<Trades trades={[]} />);
  
    expect(component).toMatchSnapshot();
  });
});
