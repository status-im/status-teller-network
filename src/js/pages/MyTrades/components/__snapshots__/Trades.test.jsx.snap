// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Trades should render correctly 1`] = `
<div
  className="mb-5"
>
  <p
    className="text-small text-muted mb-1 clickable"
    onClick={[Function]}
  >
    Show Filters 
    <FontAwesomeIcon
      border={false}
      className=""
      fixedWidth={false}
      flip={null}
      icon={
        Object {
          "icon": Array [
            320,
            512,
            Array [],
            "f0d7",
            "M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z",
          ],
          "iconName": "caret-down",
          "prefix": "fas",
        }
      }
      inverse={false}
      listItem={false}
      mask={null}
      pull={null}
      pulse={false}
      rotation={null}
      size={null}
      spin={false}
      symbol={false}
      title=""
      transform={null}
    />
  </p>
  <Card
    body={true}
    className="profile-trades-list border-0 pt-0"
    tag="div"
  >
    <Row
      className="my-1 border-bottom shadow-sm p-2 mb-3"
      tag="div"
    >
      <Col
        className="align-self-center pt-3 pb-3 text-center text-muted"
        tag="div"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
      />
    </Row>
  </Card>
</div>
`;

exports[`Trades should render when empty 1`] = `
<div
  className="mb-5"
>
  <Card
    body={true}
    className="text-center border-0 text-muted shadow-sm"
    tag="div"
  />
</div>
`;
